pkg/distributor/ha_tracker.go

```diff
@@ -204,10 +204,74 @@ func newHATracker(cfg HATrackerConfig, limits haTrackerLimits, reg prometheus.Re
 		t.client = client
 	}
 
-	t.Service = services.NewBasicService(nil, t.loop, nil)
+	t.Service = services.NewBasicService(t.starting, t.loop, nil)
 	return t, nil
 }
 
+func (h *haTracker) starting(ctx context.Context) (err error) {
+	if !h.cfg.EnableHATracker {
+		return nil
+	}
+
+	keys, err := h.client.List(ctx, "")
+	if err != nil {
+		level.Warn(h.logger).Log("msg", "starting: failed to list replica keys", "err", err)
+		return err
+	}
+
+	for _, key := range keys {
```

Comment:
I see that most parts of this loop consist of exactly same logic from the function passed to `WatchPrefix` [here](https://github.com/grafana/mimir/blob/9af647e73b2e75cdffc054770850317dac826697/pkg/distributor/ha_tracker.go#L230). It has different return type but I believe we can still extract the deduplication out and reuse the logic?